qtmodern/windows.py

```diff
@@ -79,56 +139,155 @@ def setupUi(self):
 
-        if PLATFORM == "Darwin":
-            self.hboxTitle.addWidget(self.btnClose)
-            self.hboxTitle.addWidget(self.btnMinimize)
-            self.hboxTitle.addWidget(self.btnRestore)
-            self.hboxTitle.addWidget(self.btnMaximize)
-            self.hboxTitle.addWidget(self.lblTitle)
+        allButtons = [self.btnClose, self.btnMinimize, self.btnMaximize, self.btnRestore]
+        if True in [b.isEnabled() for b in allButtons]:
+            for b in allButtons:
+                b.setVisible(True)
+            if maximized:
+                self.btnMaximize.setVisible(False)
+            else:
+                self.btnRestore.setVisible(False)
+            self.lblTitle.setContentsMargins(0, 0, 0, 0)
         else:
-            self.hboxTitle.addWidget(self.lblTitle)
-            self.hboxTitle.addWidget(self.btnMinimize)
-            self.hboxTitle.addWidget(self.btnRestore)
-            self.hboxTitle.addWidget(self.btnMaximize)
-            self.hboxTitle.addWidget(self.btnClose)
+            for b in allButtons:
+                b.setVisible(False)
+            self.lblTitle.setContentsMargins(0, 2, 0, 0)
+
+    def setWindowFlag(self, Qt_WindowType, on=True):
+        buttonHints = [Qt.WindowCloseButtonHint, Qt.WindowMinimizeButtonHint, Qt.WindowMaximizeButtonHint]
+
+        if Qt_WindowType in buttonHints:
+            self._setWindowButtonState(Qt_WindowType, on)
+        else:
+            QWidget.setWindowFlag(self, Qt_WindowType, on)
+
+    def setWindowFlags(self, Qt_WindowFlags):
+        buttonHints = [Qt.WindowCloseButtonHint, Qt.WindowMinimizeButtonHint, Qt.WindowMaximizeButtonHint]
+        for hint in buttonHints:
+            self._setWindowButtonState(hint, bool(Qt_WindowFlags & hint))
+
+        QWidget.setWindowFlags(self, Qt_WindowFlags)
+
+    @Slot()
+    def on_btnMinimize_clicked(self):
+        self.setWindowState(Qt.WindowMinimized)
+
+    @Slot()
+    def on_btnRestore_clicked(self):
+        if self.btnMaximize.isEnabled() or self.btnRestore.isEnabled():
+            self.btnRestore.setVisible(False)
+            self.btnRestore.setEnabled(False)
+            self.btnMaximize.setVisible(True)
+            self.btnMaximize.setEnabled(True)
+
+        self.setWindowState(Qt.WindowNoState)
+
+    @Slot()
+    def on_btnMaximize_clicked(self):
+        if self.btnMaximize.isEnabled() or self.btnRestore.isEnabled():
+            self.btnRestore.setVisible(True)
+            self.btnRestore.setEnabled(True)
+            self.btnMaximize.setVisible(False)
+            self.btnMaximize.setEnabled(False)
+
+        self.setWindowState(Qt.WindowMaximized)
+
+    @Slot()
+    def on_btnClose_clicked(self):
+        self.close()
+
+    @Slot()
+    def on_titleBar_doubleClicked(self):
+        if not bool(self.windowState() & Qt.WindowMaximized):
+            self.on_btnMaximize_clicked()
+        else:
+            self.on_btnRestore_clicked()
+
+
+class ModernDialog(QDialog):
```

Comment:
This shares a lot of code with ModernWindow, maybe all of it? I think it would be a good idea to break it into a `ModernWidget` which inherits `QWidget` that then contains the shared code from ModernWindow and ModernDialog. Window and Dialog can then inherit the `ModernWidget` istead of `QWidget`.